centreon/src/Core/ResourceAccess/Application/UseCase/UpdateRule/UpdateRule.php

```diff
@@ -0,0 +1,483 @@
+            $datasetFilter = null;
+            $validateAndBuildDatasetFilter($dataset, $datasetFilter);
+
+            /** @var DatasetFilter $datasetFilter */
+            $datasetFilters[] = $datasetFilter;
+        }
+
+        return $datasetFilters;
+    }
+
+    /**
+     * @param DatasetFilter $filter
+     *
+     * @return DatasetFilter
+     */
+    private function findLastLevelOfFilter(DatasetFilter $filter): DatasetFilter
+    {
+        $applicableFilter = null;
+        // recursive method to find the last 'stage' of filter (descending filter)
+        $findApplicableFilter = function (DatasetFilter $filter) use (&$findApplicableFilter, &$applicableFilter): DatasetFilter {
+            // initialize the $applicableFilter which is initially NULL
+            if ($applicableFilter === null) {
+                $applicableFilter = $filter;
+                $findApplicableFilter($applicableFilter);
+            }
+            // if there is a level then keep digging
+            elseif ($filter->getDatasetFilter() !== null) {
+                $applicableFilter = $filter->getDatasetFilter();
+                $findApplicableFilter($applicableFilter);
+            }
+
+            return $applicableFilter;
+        };
+
+        return $findApplicableFilter($filter);
+    }
+
+    /**
+     * @param int $ruleId
+     * @param int $datasetId
+     * @param DatasetFilter $filter
+     */
+    private function saveDatasetFiltersHierarchy(int $ruleId, int $datasetId, DatasetFilter $filter): void
+    {
+        $parentFilterId = null;
+
+        $saveDatasetFiltersHierarchy = function (
+            int $ruleId,
+            int $datasetId,
+            DatasetFilter $filter
+        ) use (&$parentFilterId, &$saveDatasetFiltersHierarchy): void {
+            // First iteration we save the root filter
+            $this->debug(
+                'Add dataset filter',
+                [
+                    'type' => $filter->getType(),
+                    'resource_ids' => $filter->getResourceIds(),
+                ]
+            );
+            $parentFilterId = $this->writeRepository->addDatasetFilter($ruleId, $datasetId, $filter, $parentFilterId);
+
+            // if there is a next level then save next level until final level reached
+            if ($filter->getDatasetFilter() !== null) {
+                $saveDatasetFiltersHierarchy($ruleId, $datasetId, $filter->getDatasetFilter());
+            }
+        };
+
+        $saveDatasetFiltersHierarchy($ruleId, $datasetId, $filter);
+    }
+
+    /**
+     * @param UpdateRuleRequest $updateRequest
+     */
+    private function updateResourceLinks(UpdateRuleRequest $updateRequest): void
+    {
+        // validate the updated datasets sent before doing anything
+        $this->debug('Validating updated datasets', $updateRequest->datasetFilters);
+        $datasetFilters = $this->validateAndCreateDatasetFiltersFromRequest($updateRequest);
```

Comment:
Should it be possible to move the validation part outside of your 3 methods ?

updateLinkedContacts
updateLinkedContactGroups
updateResourceLinks

If it is possible I Think it should be more readable and maintanable to validate all your data, then when everything is ok do the update

Reply:
I've though about it. But it means to check 2 times if the data needs to be updated. There is no point in validating / updating the data if the data is the exact same thing.